Add PvT-v2 Model

FoamoftheSea opened the pull request:
## Description
- **Motivation** - The PvT is a useful backbone for computer vision tasks, but only the outdated v1 is available in Hugging Face.
- **What this PR Does** - Full integration of PvT-v2 model (works with AutoModel and AutoBackbone).
- **Notes** - Like the original implementation, the config allows for using either Spatial Reduction "SR" or average pooling "AP" to reduce complexity in the attention layer, default is using SRA (as in the original code).

@amyeroberts 

## Resources
**Model paper**
- [PVT v2: Improved Baselines with Pyramid Vision Transformer](https://arxiv.org/abs/2106.13797)

**Open Source Implementations**
- [Original](https://github.com/whai362/PVT/blob/v2/classification/pvt_v2.py)
- [Panoptic Segformer](https://github.com/zhiqi-li/Panoptic-SegFormer/blob/master/easymd/models/backbones/pvt_v2.py)

## Checks
- Add PvT-v2 to model docs ✔
- Build pytests and have them pass ✔
- Formatting (make fix-copies, make fixup) ✔
- Convert open-source weights and test expected logits, uploaded to hub ✔

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
Allowed for batching of eval metrics

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
copied models/pvt to adapt to pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
First commit of pvt_v2

FoamoftheSea committed:
Set key and value layers to use separate linear modules. Fixed pruning function

FoamoftheSea committed:
Set key and value layers to use separate linear modules. Fixed pruning function

FoamoftheSea committed:
Set key and value layers to use separate linear modules. Fixed pruning function

FoamoftheSea committed:
Set key and value layers to use separate linear modules. Fixed pruning function

FoamoftheSea committed:
Set key and value layers to use separate linear modules. Fixed pruning function

FoamoftheSea committed:
Set AvgPool to 7

FoamoftheSea committed:
Set AvgPool to 7

FoamoftheSea committed:
Set AvgPool to 7

FoamoftheSea committed:
Set AvgPool to 7

FoamoftheSea committed:
Set AvgPool to 7

FoamoftheSea committed:
Fixed issue in init

FoamoftheSea committed:
Fixed issue in init

FoamoftheSea committed:
Fixed issue in init

FoamoftheSea committed:
Fixed issue in init

FoamoftheSea committed:
Fixed issue in init

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
PvT-v2 now works in AutoModel

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2 models

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2 models

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2 models

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2 models

FoamoftheSea committed:
Successful conversion of pretrained weights for PVT-v2 models

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pytests for pvt-v2, all passed

FoamoftheSea committed:
Added pvt_v2 to docs/source/end/model_doc

FoamoftheSea committed:
Added pvt_v2 to docs/source/end/model_doc

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Ran fix-copies and fixup. All checks passed

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
Added additional ReLU for linear attention mode

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
pvt_v2_b2_linear converted and working

FoamoftheSea committed:
Reverted batch eval changes for PR

FoamoftheSea committed:
Reverted batch eval changes for PR

FoamoftheSea committed:
Reverted batch eval changes for PR

FoamoftheSea committed:
Reverted batch eval changes for PR

FoamoftheSea committed:
Reverted batch eval changes for PR

FoamoftheSea committed:
Updated index.md

FoamoftheSea committed:
Expanded type support for Pvt-v2 config

FoamoftheSea committed:
Expanded type support for Pvt-v2 config

FoamoftheSea committed:
Expanded type support for Pvt-v2 config

FoamoftheSea committed:
Expanded type support for Pvt-v2 config

FoamoftheSea committed:
Fixed config docstring. Added channels property

FoamoftheSea committed:
Fixed config docstring. Added channels property

FoamoftheSea committed:
Fixed config docstring. Added channels property

FoamoftheSea committed:
Fixed config docstring. Added channels property

FoamoftheSea committed:
Fixed config docstring. Added channels property

ArthurZucker commented:
FYI @rafaelpadilla

FoamoftheSea committed:
Fixed model names in tests

FoamoftheSea committed:
Fixed model names in tests

FoamoftheSea committed:
Fixed model names in tests

FoamoftheSea committed:
Fixed model names in tests

amyeroberts commented:
@FoamoftheSea - awesome work! Let us know when you're ready for review. 

For the code quality checks, running `make fixup` and pushing the changes should resolve them.

FoamoftheSea commented:
@amyeroberts  - Thanks! Let me try that and do one final sweep over things, then I will get back to you shortly for review :)

FoamoftheSea committed:
Fixed config backbone compat. Added additional type support for image size in config

FoamoftheSea committed:
Fixed config backbone compat. Added additional type support for image size in config

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Fixed config backbone compat

FoamoftheSea committed:
Ran fix-copies

FoamoftheSea committed:
Ran fix-copies

FoamoftheSea committed:
Ran fix-copies

FoamoftheSea committed:
Fixed PvtV2Backbone tests

FoamoftheSea committed:
Fixed PvtV2Backbone tests

FoamoftheSea committed:
Added TFRegNet to OBJECTS_TO_IGNORE in check_docstrings.py

FoamoftheSea committed:
Added TFRegNet to OBJECTS_TO_IGNORE in check_docstrings.py

FoamoftheSea committed:
Fixed backbone stuff and fixed tests: all passing

FoamoftheSea committed:
Fixed backbone stuff and fixed tests: all passing

FoamoftheSea committed:
Ran make fixup

FoamoftheSea committed:
Ran make fixup

FoamoftheSea committed:
Made modifications for code checks

FoamoftheSea committed:
Made modifications for code checks

FoamoftheSea commented:
@amyeroberts I believe this is ready for review now. I made changes so all the quality checks pass, and I also added integration with AutoBackbone so that the PVTv2 can be used with Deformable DETR and other models that use AutoBackbone.

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+    ):
+        super().__init__()
+        self.fp16_enabled = False
+        patch_size = _ntuple(2)(patch_size)
```
For consistency with the rest of the repo 

```suggestion
        patch_size = (patch_size, patch_size) if isinstance(patch_size, int) else patch_size
```

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+class PvtV2OverlapPatchEmbeddings(nn.Module):
+    """Image to Patch Embedding"""
+
+    def __init__(
```
Should still take the config as its first argument

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+        hidden_size: int,
+    ):
+        super().__init__()
+        self.fp16_enabled = False
```
What is this for?

FoamoftheSea replied:
This was copied in by mistake, it's removed now

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+
+
+class PvTV2DWConv(nn.Module):
+    def __init__(self, dim=768):
```
Should take `config` as its first argument

amyeroberts replied:
Should still take the `config` as its first argument

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+        return embeddings, height, width
+
+
+class PvTV2DWConv(nn.Module):
```
Could you add a docstring here to detail what the abbreviated name means i.e. what is "dw"?

amyeroberts reviewed (commented):
Really nice PR - thanks for adding! 

Just a few small comments - the main one is that we don't need the image processor. Once resolved we should be good to merge!

FoamoftheSea committed:
Remove ONNX config from configuration_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Remove ONNX config from configuration_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Use explicit image size dict in test_modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Use explicit image size dict in test_modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Make image_size optional in test_modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Make image_size optional in test_modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea commented on utils/check_docstrings.py:
```diff
@@ -653,6 +653,8 @@
     "TFRagModel",
     "TFRagSequenceForGeneration",
     "TFRagTokenForGeneration",
+    "TFRegNetModel",
+    "TFRegNetForImageClassification",
```
Agreed. These lines were added because their omission was causing this check to fail when running make fixup. I'm not sure how that check is passing elsewhere without this, but we can remove these changes from this PR is that's more appropriate.

FoamoftheSea replied:
```suggestion
```

FoamoftheSea replied:
@amyeroberts I ran some tests and I think everything is ready to go except for this last change. The checks have passed and the gradient checkpointing is working now, so with this last commit I think it may be ready for merge, let me know what you think.

FoamoftheSea committed:
Remove _ntuple use in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Remove _ntuple use in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea commented on src/transformers/models/pvt_v2/image_processing_pvt_v2.py:
```diff
@@ -0,0 +1,268 @@
+logger = logging.get_logger(__name__)
+
+
+class PvtV2ImageProcessor(BaseImageProcessor):
```
Agreed, although it's even closer to duplicating PvtImageProcessor which uses the same default mean/std and already in the repo, so should we use that instead?

amyeroberts replied:
Yes please!

FoamoftheSea committed:
Remove reference to fp16_enabled

FoamoftheSea committed:
Remove reference to fp16_enabled

FoamoftheSea committed:
Model modules now take config as first argument even when not used

FoamoftheSea committed:
Model modules now take config as first argument even when not used

FoamoftheSea committed:
Replaced abbreviations for "SR" and "AP" with explicit "spatialreduction" and "averagepooling"

FoamoftheSea committed:
Replaced abbreviations for "SR" and "AP" with explicit "spatialreduction" and "averagepooling"

FoamoftheSea committed:
All LayerNorm now instantiates with config.layer_norm_eps

FoamoftheSea committed:
All LayerNorm now instantiates with config.layer_norm_eps

FoamoftheSea committed:
Added docstring for depth-wise conv layer

FoamoftheSea committed:
Added docstring for depth-wise conv layer

FoamoftheSea committed:
PvtV2Config now only takes Union[int, Tuple[int, int]] for image size

FoamoftheSea committed:
PvtV2Config now only takes Union[int, Tuple[int, int]] for image size

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+
+        hidden_states = outputs.hidden_states
+
+        feature_maps = ()
+        for idx, stage in enumerate(self.stage_names):
+            if stage in self.out_features:
+                feature_maps += (hidden_states[idx],)
```
^---- this is now obsolete

FoamoftheSea committed:
Refactored PVTv2 in prep for gradient checkpointing

FoamoftheSea committed:
Refactored PVTv2 in prep for gradient checkpointing

FoamoftheSea committed:
Gradient checkpointing ready to test

FoamoftheSea committed:
Gradient checkpointing ready to test

FoamoftheSea committed:
Removed override of _set_gradient_checkpointing

FoamoftheSea committed:
Removed override of _set_gradient_checkpointing

FoamoftheSea committed:
Cleaned out old code

FoamoftheSea committed:
Cleaned out old code

FoamoftheSea committed:
Applied code fixup

FoamoftheSea committed:
Applied code fixup

FoamoftheSea committed:
Applied code fixup

FoamoftheSea committed:
Applied code fixup

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,716 @@
+
+        batch_size = pixel_values.shape[0]
+        hidden_states = pixel_values
+        for idx, x in enumerate(zip(self.patch_embeddings, self.block, self.layer_norms)):
```
Added ✔️

FoamoftheSea replied:
Hey @amyrobers - in the recent commits I restructured the model a bit and added the gradient checkpointing logic, but although it runs it gives me this warning:
```
UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None"
```
I ran a test and the weights in the encoder do not train when gradient checkpointing is enabled, so this is not implemented properly. Is there something special that I need to do to get it to work?

FoamoftheSea replied:
Hey @amyeroberts - in the recent commits I restructured the model a bit and added the gradient checkpointing logic, but although it runs it gives me this warning:
```
UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None"
```
I ran a test and the weights in the encoder do not train when gradient checkpointing is enabled, so this is not implemented properly. Is there something special that I need to do to get it to work?

amyeroberts replied:
@FoamoftheSea 🤔 interesting! Could you share snippet you're using to test the model?

FoamoftheSea replied:
Sure thing, if you run this script after pulling the latest changes with its default settings it should download and unzip the tiny ImageNet dataset and begin a short training.

```python
from argparse import ArgumentParser
from collections import defaultdict
from pathlib import Path

import imageio

import numpy as np

import torch
from torch.utils.data import Dataset

import os
import urllib
import zipfile

from tqdm import tqdm

from transformers import Trainer, TrainingArguments, PvtV2ForImageClassification, PvtV2Config

use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")


def download_dataset(path):
    print('Beginning dataset download with urllib2')
    url = "http://cs231n.stanford.edu/tiny-imagenet-200.zip"
    urllib.request.urlretrieve(url, "%s/tiny-imagenet-200.zip" % path)
    print("Dataset downloaded")


def unzip_data(path):
    path_to_zip_file = "%s/tiny-imagenet-200.zip" % path
    directory_to_extract_to = path
    print("Extracting zip file: %s" % path_to_zip_file)
    with zipfile.ZipFile(path_to_zip_file, 'r') as zip_ref:
        zip_ref.extractall(directory_to_extract_to)
    print("Extracted at: %s" % directory_to_extract_to)


def download_and_unzip(URL, root_dir):
    path = Path(os.path.abspath(root_dir))
    path.mkdir(parents=True, exist_ok=True)
    download_dataset(path)
    unzip_data(path)

    root_dir = "%s/tiny-imagenet-200" % str(path)

    return root_dir


class TinyImageNetPaths:
    def __init__(self, root_dir, download=False):
        if download:
            root_dir = download_and_unzip('http://cs231n.stanford.edu/tiny-imagenet-200.zip', root_dir)
        train_path = os.path.join(root_dir, 'train')
        val_path = os.path.join(root_dir, 'val')
        test_path = os.path.join(root_dir, 'test')

        wnids_path = os.path.join(root_dir, 'wnids.txt')
        words_path = os.path.join(root_dir, 'words.txt')

        self._make_paths(train_path, val_path, test_path,
                         wnids_path, words_path)

    def _make_paths(self, train_path, val_path, test_path,
                    wnids_path, words_path):
        self.ids = []
        with open(wnids_path, 'r') as idf:
            for nid in idf:
                nid = nid.strip()
                self.ids.append(nid)
        self.nid_to_words = defaultdict(list)
        with open(words_path, 'r') as wf:
            for line in wf:
                nid, labels = line.split('\t')
                labels = list(map(lambda x: x.strip(), labels.split(',')))
                self.nid_to_words[nid].extend(labels)

        self.paths = {'train': [], 'val': [], 'test': list(map(lambda x: os.path.join(test_path, x),
                                                               os.listdir(test_path)))}

        # Get the test paths
        # Get the validation paths and labels
        with open(os.path.join(val_path, 'val_annotations.txt')) as valf:
            for line in valf:
                fname, nid, x0, y0, x1, y1 = line.split()
                fname = os.path.join(val_path, 'images', fname)
                bbox = int(x0), int(y0), int(x1), int(y1)
                label_id = self.ids.index(nid)
                self.paths['val'].append((fname, label_id, nid, bbox))

        # Get the training paths
        train_nids = os.listdir(train_path)
        for nid in train_nids:
            anno_path = os.path.join(train_path, nid, nid + '_boxes.txt')
            imgs_path = os.path.join(train_path, nid, 'images')
            label_id = self.ids.index(nid)
            with open(anno_path, 'r') as annof:
                for line in annof:
                    fname, x0, y0, x1, y1 = line.split()
                    fname = os.path.join(imgs_path, fname)
                    bbox = int(x0), int(y0), int(x1), int(y1)
                    self.paths['train'].append((fname, label_id, nid, bbox))


def _add_channels(img, total_channels=3):
    while len(img.shape) < 3:  # third axis is the channels
        img = np.expand_dims(img, axis=-1)
    while (img.shape[-1]) < 3:
        img = np.concatenate([img, img[:, :, -1:]], axis=-1)
    return img


class TinyImageNetDataset(Dataset):
    def __init__(self, root_dir, mode='train', preload=True, load_transform=None,
                 transform=None, download=False, max_samples=None):
        tinp = TinyImageNetPaths(root_dir, download)
        self.mode = mode
        self.label_idx = 1  # from [image, id, nid, box]
        self.preload = preload
        self.transform = transform
        self.transform_results = dict()

        self.IMAGE_SHAPE = (64, 64, 3)

        self.img_data = []
        self.label_data = []

        self.max_samples = max_samples
        self.samples = tinp.paths[mode]
        self.samples_num = len(self.samples)

        if self.max_samples is not None:
            self.samples_num = min(self.max_samples, self.samples_num)
            np.random.shuffle(self.samples)
            self.samples = self.samples[:self.samples_num]

        if self.preload:
            load_desc = "Preloading {} data...".format(mode)
            self.img_data = np.zeros((self.samples_num,) + self.IMAGE_SHAPE,
                                     dtype=np.float32)
            self.label_data = np.zeros((self.samples_num,), dtype=np.int32)
            for idx in tqdm(range(self.samples_num), desc=load_desc):
                s = self.samples[idx]
                if not isinstance(s, tuple):
                    continue
                img = imageio.imread(s[0])
                img = _add_channels(img)
                self.img_data[idx] = img
                if mode != 'test':
                    self.label_data[idx] = s[self.label_idx]

            if load_transform:
                for lt in load_transform:
                    result = lt(self.img_data, self.label_data)
                    self.img_data, self.label_data = result[:2]
                    if len(result) > 2:
                        self.transform_results.update(result[2])

    def __len__(self):
        return self.samples_num

    def __getitem__(self, idx):
        if self.preload:
            img = self.img_data[idx]
            lbl = None if self.mode == 'test' else self.label_data[idx]
        else:
            s = self.samples[idx]
            img = imageio.imread(s[0])
            lbl = None if self.mode == 'test' else s[self.label_idx]
        lbl_oh = np.zeros(200)
        lbl_oh[lbl] = 1
        sample = {'pixel_values': img.transpose(2, 0, 1), 'label': lbl_oh}

        if self.transform:
            sample = self.transform(sample)
        return sample


def main(args):
    # Create DataLoaders for pre-trained models (normalized based on specific requirements)
    download = False
    try:
        train_loader_pretrain = TinyImageNetDataset(root_dir=args.data_root, mode="train", max_samples=5, download=download)
    except FileNotFoundError:
        download = True
        train_loader_pretrain = TinyImageNetDataset(root_dir=args.data_root, mode="train", max_samples=5, download=download)

    data_root = os.path.join(args.data_root, "tiny-imagenet-200") if download else args.data_root
    val_loader_pretrain = TinyImageNetDataset(root_dir=data_root, mode="test", max_samples=5)

    model = PvtV2ForImageClassification.from_pretrained(
        "FoamoftheSea/pvt_v2_b0",
        config=PvtV2Config(num_labels=200),
        ignore_mismatched_sizes=True
    )

    training_args = TrainingArguments(
        "pvt-v2-test",
        num_train_epochs=10,
        per_device_train_batch_size=args.batch_size,
        per_device_eval_batch_size=args.batch_size,
        save_total_limit=3,
        evaluation_strategy="steps",
        save_strategy="steps",
        save_steps=10000,
        eval_steps=1000,
        logging_steps=1,
        load_best_model_at_end=True,
        push_to_hub=False,
        gradient_checkpointing=True,
        use_cpu=args.use_cpu,
    )

    trainer = Trainer(model, args=training_args, train_dataset=train_loader_pretrain, eval_dataset=val_loader_pretrain)

    trainer.train()


if __name__ == "__main__":
    parser = ArgumentParser(description="Test PvtV2 on Tiny ImageNet.")
    parser.add_argument("-d", "--data-root", default="./tiny-imagenet-200")
    parser.add_argument("-cpu", "--use-cpu", action="store_true", default=False)
    parser.add_argument("-bs", "--batch-size", type=int, default=64)

    args = parser.parse_args()

    main(args)
```

amyeroberts replied:
Thanks for sharing this! There's a related issue here: https://github.com/huggingface/transformers/issues/26969#issuecomment-1807831645

If you pass in `gradient_checkpointing_kwargs={'use_reentrant':False}` to `TrainingArguments` it should now work.

FoamoftheSea replied:
I tested this and it works using that option. I adjusted the model tests to reflect this.

FoamoftheSea committed:
Merge remote-tracking branch 'origin/pvtv2' into pvtv2

# Conflicts:
#	src/transformers/__init__.py
#	src/transformers/models/pvt_v2/__init__.py

FoamoftheSea committed:
Began debug of pvt_v2 tests

FoamoftheSea committed:
Began debug of pvt_v2 tests

FoamoftheSea committed:
Leave handling of num_labels to base pretrained config class

FoamoftheSea committed:
Leave handling of num_labels to base pretrained config class

FoamoftheSea committed:
Merge remote-tracking branch 'origin/pvtv2' into pvtv2

FoamoftheSea committed:
Deactivated gradient checkpointing tests until it is fixed

FoamoftheSea committed:
Deactivated gradient checkpointing tests until it is fixed

FoamoftheSea committed:
Removed PvtV2ImageProcessor which duped PvtImageProcessor

FoamoftheSea committed:
Removed PvtV2ImageProcessor which duped PvtImageProcessor

FoamoftheSea committed:
Merge remote-tracking branch 'origin/pvtv2' into pvtv2

# Conflicts:
#	src/transformers/models/pvt_v2/modeling_pvt_v2.py

FoamoftheSea committed:
Fixed issue from rebase

FoamoftheSea committed:
Fixed issue from rebase

FoamoftheSea committed:
Set tests for gradient checkpointing to skip those using reentrant since it isn't supported

FoamoftheSea committed:
Fixed issue from rebase

FoamoftheSea committed:
Fixed issue from rebase

FoamoftheSea committed:
Changed model name in docs

FoamoftheSea committed:
Removed duplicate PvtV2Backbone

FoamoftheSea committed:
Merge pull request #17 from FoamoftheSea/pvtv2

Pvtv2

FoamoftheSea commented on docs/source/en/index.md:
```diff
@@ -48,8 +48,236 @@ The documentation is organized into five sections:
   - **MODELS** details the classes and functions related to each model implemented in the library.
   - **INTERNAL HELPERS** details utility classes and functions used internally.
 
+### Supported models
```
Fixed ✔️

FoamoftheSea committed:
Work around type switching issue in tests

FoamoftheSea committed:
Merge branch 'pvtv2' into pvtv2-staging

FoamoftheSea committed:
Fix model name in config comments

FoamoftheSea commented:
Hey @amyeroberts - just adding a new comment here since the others are pretty well buried. This PR should be ready to go after removing the two lines from the check_docstrings.py. Let me know if you agree and we can make that last commit and merge 🙏

FoamoftheSea commented:
Great, thanks @ArthurZucker! Let me know if there's any action needed from my side, do you want me to address the failure from the documentation build?

ArthurZucker commented:
Sorry I started reviewing and did not finish yet

ArthurZucker reviewed (commented):
Overall looks good to me, left a few nits here and there!

FoamoftheSea committed:
Update docs/source/en/model_doc/pvt_v2.md

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+            hidden_states = outputs[0]
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (outputs[1],)
+            # optionally reshape back to (batch_size, num_channels, height, width)
+            if idx != len(self.layers) - 1 or (idx == len(self.layers) - 1 and self.config.reshape_last_stage):
+                hidden_states = hidden_states.reshape(batch_size, height, width, -1).permute(0, 3, 1, 2).contiguous()
```
This would be useful if someone is turning the image into a token rather than a feature map. I got this from the Segformer model in this same repo which has a nearly identical base, I would vote to keep it in.

amyeroberts replied:
I don't see why the user can't just do that themselves with the model output after the forward pass?

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        # Patch embedding
+        self.patch_embedding = PvtV2OverlapPatchEmbeddings(
+            config=config,
+            patch_size=config.patch_sizes[layer_idx],
+            stride=config.strides[layer_idx],
+            num_channels=config.num_channels if layer_idx == 0 else config.hidden_sizes[layer_idx - 1],
+            hidden_size=config.hidden_sizes[layer_idx],
```
Sorry I don't understand, can you explain or show me what you mean in a code suggestion? This is the number of input channels for the 2D conv, so it should align with the hidden size of the previous layer, or the number of channels of the input image for the very first layer, so this looks right to me.

FoamoftheSea replied:
I've implemented your suggestion in https://github.com/huggingface/transformers/pull/26812/commits/f3109f005c47fc7000c9d83dab1c62542bb371e3

FoamoftheSea commented on src/transformers/models/pvt_v2/convert_pvt_v2_to_pytorch.py:
```diff
@@ -0,0 +1,279 @@
+        expected_slice_logits = torch.tensor([-0.1769, -0.1747, -0.0143])
+    elif pvt_v2_size == "b5":
+        expected_slice_logits = torch.tensor([-0.2943, -0.1008, 0.6812])
+    else:
```
This is specifically for converting the original published ImageNet weights of the PVTv2, and also a hard check to make sure those weights have been copied correctly. Both the PVT and Segformer model in this repository also perform this check, which is where I took the code from.

FoamoftheSea replied:
ImageNet weight verification is now opt-in with `--verify-imagenet-weights` option, since I suppose users are far more likely to use this for their own models, since the ImageNet weights are available on the hub.

Commit with fix: https://github.com/huggingface/transformers/pull/26812/commits/d33f0eb492780dd8f1b5519866d1ad2754e81211

FoamoftheSea commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -176,6 +176,7 @@
         ("pop2piano", "Pop2PianoConfig"),
         ("prophetnet", "ProphetNetConfig"),
         ("pvt", "PvtConfig"),
+        ("pvt_v2", "PvtV2Config"),
```
The naming format is mixed in this file, for instance the mobilenet_v1 and mobilenet_v2 use snake case. Another popular implementation of the PVTv2 model is in the timm library where is is referred to as pvt_v2, and also that's how it appears in the file tree, so I thought snake case seemed more appropriate. Is there a hard expectation to hyphenate this?

FoamoftheSea replied:
Revisiting this, since the most commonly used implementation of pvt_v2 is currently in the timm library under that name, I think people would appreciate consistency between these libraries. What do you think?

amyeroberts replied:
OK - let's go for the underscore for consistency with timm here

FoamoftheSea committed:
Changed name of variable from 'attn_reduce' to 'sr_type'

FoamoftheSea committed:
Changed name of variable from 'attn_reduce' to 'sr_type'

FoamoftheSea committed:
Changed from using 'sr_type' to 'linear_attention' for clarity

FoamoftheSea committed:
Merge branch 'pvtv2' into pvtv2-staging

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+    def __init__(self, config: PvtV2Config):
+        super().__init__(config)
+        self.config = config
+        # self.channels = [config.hidden_sizes[i] for i in config.out_indices]
```
```suggestion
```

FoamoftheSea committed:
Update src/transformers/models/pvt_v2/modeling_pvt_v2.py

Removed old code

FoamoftheSea commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,175 @@
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether or not a learnable bias should be added to the queries, keys and values.
+        attn_reduce (`str`, *optional*, defaults to `"spatialreduction"`):
+            Attention complexity reduction type. Choice of 'spatialreduction' or 'averagepooling' .
```
Resolved with conversion of this variable to a boolean instead of string.

FoamoftheSea committed:
Changed from using 'sr_type' to 'linear_attention' for clarity

FoamoftheSea committed:
Fixed Class names to be more descriptive

FoamoftheSea committed:
Merge branch 'pvtv2' into pvtv2-staging

FoamoftheSea committed:
Merge remote-tracking branch 'origin/pvtv2-staging' into pvtv2-staging

FoamoftheSea commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,175 @@
+            The epsilon used by the layer normalization layers.
+        qkv_bias (`bool`, *optional*, defaults to `True`):
+            Whether or not a learnable bias should be added to the queries, keys and values.
+        attn_reduce (`str`, *optional*, defaults to `"spatialreduction"`):
```
Resolved by converting this parameter to a boolean called "linear_attention" which is how the [original code implementation](https://github.com/whai362/PVT/blob/57e2dfaa5a46f9050d76f306a4fcd9a7c061f520/classification/pvt_v2.py#L395) handles this option.

FoamoftheSea committed:
Merge remote-tracking branch 'origin/pvtv2' into pvtv2

FoamoftheSea committed:
Merge branch 'pvtv2' into pvtv2-staging

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        self.pruned_heads = self.pruned_heads.union(heads)
+
+
+class PvtV2ConvFFN(nn.Module):
```
Fixed in https://github.com/huggingface/transformers/pull/26812/commits/680be579f2ab6f889c7850f2265b896d701ae081

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        return embeddings, height, width
+
+
+class PvTV2DWConv(nn.Module):
```
Fixed in https://github.com/huggingface/transformers/pull/26812/commits/680be579f2ab6f889c7850f2265b896d701ae081

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        # This is actually dropping out entire tokens to attend to, which might
+        # seem a bit unusual, but is taken from the original Transformer paper.
+        attention_probs = self.attn_drop(attention_probs)
+        # context_layer = context_layer.permute(0, 2, 1, 3).contiguous()
+        # new_context_layer_shape = context_layer.size()[:-2] + (self.all_head_size,)
+        # context_layer = context_layer.view(new_context_layer_shape)
```
Yes
```suggestion
```

FoamoftheSea committed:
Update src/transformers/models/pvt_v2/modeling_pvt_v2.py

Removed outdated code

FoamoftheSea committed:
Moved paper abstract to single line in pvt_v2.md

FoamoftheSea committed:
Added usage tips to pvt_v2.md

FoamoftheSea commented on docs/source/en/model_doc/pvt_v2.md:
```diff
@@ -0,0 +1,71 @@
+
+Abstract from the paper:
+
+*Transformer recently has presented encouraging
```
Fixed in https://github.com/huggingface/transformers/pull/26812/commits/054b281d50f0a8fd2cf5fb5ab9c3196a81d33467

FoamoftheSea commented on docs/source/en/model_doc/pvt_v2.md:
```diff
@@ -0,0 +1,71 @@
+https://github.com/whai362/PVT.*
+
+This model was contributed by [FoamoftheSea](https://huggingface.co/FoamoftheSea). The original code can be found [here](https://github.com/whai362/PVT).
+
```
I've added a tips section in https://github.com/huggingface/transformers/pull/26812/commits/f9c6d8af4bc8ae3be05611d441c0367fb1e72795

How does that look?

FoamoftheSea replied:
I filled out the documentation to be much more comprehensive, including expanded tips section, in https://github.com/huggingface/transformers/pull/26812/commits/03d67b3f4eb6f16246e7d2ef8a6304b021d066ea

Marking this as resolved.

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        self.sr_ratio = sr_ratio
+        if self.attn_reduce == "averagepooling":
+            self.pool = nn.AdaptiveAvgPool2d(7)
+            self.sr = nn.Conv2d(self.hidden_size, self.hidden_size, kernel_size=1, stride=1)
```
This has been fixed in https://github.com/huggingface/transformers/pull/26812/commits/cf614f3d6355dd94e0f909229359519380439fdd

I chose to address this option the same way as in the original published code, where it is treated as a True/False for using linear attention complexity mode (previously "averagepooling"), so now it is a config parameter called "linear_attention"

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+        out_features = out_features if out_features is not None else in_features
+        self.dense1 = nn.Linear(in_features, hidden_features)
+        self.dwconv = PvTV2DWConv(config, hidden_features)
+        if isinstance(config.hidden_act, str):
```
Hmm, I'm seeing that the PvtV2Config docstring says this can be a string or function, am I missing something somewhere else?

FoamoftheSea replied:
I did see an inconsistency in the typing. Take a look at https://github.com/huggingface/transformers/pull/26812/commits/033934feade953b3eacb988ed33a139b3f067a14 and see if that resolves this issue for you.

FoamoftheSea committed:
Simplified module inits by passing layer_idx

FoamoftheSea committed:
Fixed typing for hidden_act in PvtV2Config

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,731 @@
+                    num_attention_heads=config.num_attention_heads[layer_idx],
+                    drop_path=drop_path_decays[sum(config.depths[:layer_idx]) + i],
+                    sr_ratio=config.sr_ratios[layer_idx],
+                    mlp_ratio=config.mlp_ratios[layer_idx],
```
I ended up passing the drop_path rate with the layer_idx in https://github.com/huggingface/transformers/pull/26812/commits/f3109f005c47fc7000c9d83dab1c62542bb371e3

This also comes out to 3 arguments, with the drop_rate able to take a default value of 0.0

FoamoftheSea committed:
Removed unusued import

FoamoftheSea commented:
@ArthurZucker Thanks for the review! I've gone through and made commits to address most of your comments, and I've responded to a couple that await your review. Let me know when you have a chance to go through the updates.

ArthurZucker commented:
Hey! Sure, can you first make sure the CIs are green? 🤗

FoamoftheSea committed:
Merge branch 'main' into pvtv2

FoamoftheSea committed:
Merge pull request #19 from FoamoftheSea/pvtv2

Merge Pvtv2 branch

ArthurZucker commented:
Thanks, will review

FoamoftheSea committed:
Add pvt_v2 to docs/source/en/_toctree.yml

FoamoftheSea committed:
Updated documentation in docs/source/en/model_doc/pvt_v2.md to be more comprehensive.

FoamoftheSea committed:
Updated documentation in docs/source/en/model_doc/pvt_v2.md to be more comprehensive.

FoamoftheSea commented:
Hey @ArthurZucker - CIs are green! I went through and made all of the adjustments, as well as filling out the documentation to be much more comprehensive. I marked all the obviously fixed conversations as "resolved", but I left a few open that are more opinion-based. Let me know what you think.

FoamoftheSea committed:
Merge branch 'main' into pvtv2-staging

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+            else:
+                layer_output = layer(hidden_states, output_attentions)
+            outputs, height, width = layer_output
+            hidden_states = outputs[0]
+            if output_attentions:
+                all_self_attentions = all_self_attentions + (outputs[1],)
```
nit - just to make it a little clearer 

```suggestion
            hidden_states, attentions = outputs[0], outputs[1]
            if output_attentions:
                all_self_attentions = all_self_attentions + (attentions,)
```

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+        layer_idx: int,
+    ):
+        super().__init__()
+        # Patch embedding
```
nit - superfluous, it's clear from the var name what it is 

```suggestion
```

amyeroberts commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,176 @@
+# coding=utf-8
+# Copyright 2023 Authors: Wenhai Wang, Enze Xie, Xiang Li, Deng-Ping Fan,
```
```suggestion
# Copyright 2024 Authors: Wenhai Wang, Enze Xie, Xiang Li, Deng-Ping Fan,
```

amyeroberts commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,176 @@
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        if "reshape_last_stage" in kwargs and kwargs["reshape_last_stage"] is False:
+            warnings.warn(
+                "Reshape_last_stage is set to False in this config. This argument is deprecated and will soon be"
+                " removed, as the behaviour will default to that of reshape_last_stage = True.",
+                FutureWarning,
+            )
```
Let's not add arguments that are about to be deprecated!

FoamoftheSea replied:
Removed the reshape_last_stage option, I agree it's not useful and should be deprecated

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+class PvtV2OverlapPatchEmbeddings(nn.Module):
+    """Image to Patch Embedding"""
+
+    def __init__(
+        self,
+        config: PvtV2Config,
+        layer_idx: int,
+    ):
```
nit

```suggestion
    def __init__(self, config: PvtV2Config, layer_idx: int):
```

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+        layer_idx: int,
+    ):
+        super().__init__()
+        patch_size: Union[int, Iterable[int]] = config.patch_sizes[layer_idx]
+        patch_size = (patch_size, patch_size) if isinstance(patch_size, int) else patch_size
+        stride: int = config.strides[layer_idx]
+        num_channels: int = config.num_channels if layer_idx == 0 else config.hidden_sizes[layer_idx - 1]
+        hidden_size: int = config.hidden_sizes[layer_idx]
```
I'd prefer for the types here to be removed - the types of the arguments are pretty obvious from the code here 

```suggestion
        patch_size = config.patch_sizes[layer_idx]
        patch_size = (patch_size, patch_size) if isinstance(patch_size, int) else patch_size
        stride = config.strides[layer_idx]
        num_channels = config.num_channels if layer_idx == 0 else config.hidden_sizes[layer_idx - 1]
        hidden_size = config.hidden_sizes[layer_idx]
```

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+        self.sr_ratio = sr_ratio
+        if self.linear_attention:
+            self.pool = nn.AdaptiveAvgPool2d(7)
+            self.sr = nn.Conv2d(self.hidden_size, self.hidden_size, kernel_size=1, stride=1)
```
Let's use a more meaningful name here 

```suggestion
            self.sequence_reduction = nn.Conv2d(self.hidden_size, self.hidden_size, kernel_size=1, stride=1)
```

amyeroberts commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+        self.proj = nn.Linear(self.hidden_size, self.hidden_size)
+        self.proj_drop = nn.Dropout(config.hidden_dropout_prob)
+
+        self.sr_ratio = sr_ratio
```
Same here

```suggestion
        self.sequence_reduction_ratio = sequence_reduction_ratio
```

or perhaps: `self.seq_reduction_ratio`

FoamoftheSea replied:
From the paper, I think that "spatial_reduction_ratio" is the best way to go and the most accurate, since it actually has a quadratic effect on the flattened sequence length. I have an incoming commit to fix this.

amyeroberts reviewed (commented):
Thanks for all your work on this @FoamoftheSea! 

Just a few small comments.

FoamoftheSea committed:
Update src/transformers/models/pvt_v2/modeling_pvt_v2.py

Move function parameters to single line

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Update src/transformers/models/pvt_v2/modeling_pvt_v2.py

Update year of copyright to 2024

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Update src/transformers/models/pvt_v2/modeling_pvt_v2.py

Make code more explicit

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Updated sr_ratio to be more explicit spatial_reduction_ratio

FoamoftheSea committed:
Removed excess type hints in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Move params to single line in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Removed needless comment in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Update copyright date in pvt_v2.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Moved params to single line in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Updated copyright date in configuration_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Cleaned comments in modeling_pvt_v2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

FoamoftheSea committed:
Renamed spatial_reduction Conv2D operation

FoamoftheSea committed:
Revert "Update src/transformers/models/pvt_v2/modeling_pvt_v2.py
"

This reverts commit c4a04416dde8f3475ab405d1feb368600e0f8538.

FoamoftheSea committed:
Updated conversion script to reflect module name change

FoamoftheSea committed:
Deprecated reshape_last_stage option in config

FoamoftheSea committed:
Merge branch 'pvtv2' into pvtv2-staging

FoamoftheSea committed:
Removed unused imports

FoamoftheSea committed:
Code formatting

FoamoftheSea commented on tests/models/pvt_v2/test_modeling_pvt_v2.py:
Does this apply only to methods that have exactly the same text, or if there's a name change from PvtConfig -> PvtV2Config in the code block does that also apply?

amyeroberts replied:
Also applies for name changes

FoamoftheSea commented on src/transformers/models/pvt_v2/modeling_pvt_v2.py:
```diff
@@ -0,0 +1,723 @@
+_IMAGE_CLASS_EXPECTED_OUTPUT = "LABEL_281"  # ImageNet ID for "tabby, tabby cat"
+
+PVT_V2_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "FoamoftheSea/pvt_v2_b0",
```
@amyeroberts Understood, do you want me to try to reach out to them?

amyeroberts replied:
Yes please!

FoamoftheSea replied:
Tried to reach out via email, will let you know if I get a reply there. @whai362

FoamoftheSea replied:
Hey @amyeroberts - I got a response this morning from Wenhai, they asked if https://huggingface.co/OpenGVLab would work?

amyeroberts replied:
Great! Yep, that works.

amyeroberts replied:
I'll do a final review and once done and all weights stable I can help move them across

FoamoftheSea replied:
Awesome. I should have time tomorrow to complete the finishing touches here and hand back over to you for review. Will let you know!

FoamoftheSea committed:
Fixed outdated decorators on test_inference_fp16

FoamoftheSea committed:
Added "Copied from" comments in test_modeling_pvt_v2.py

FoamoftheSea committed:
Merge branch 'main' into pvtv2

# Conflicts:
#	README_hd.md
#	docs/source/en/tasks/image_classification.md

FoamoftheSea committed:
Fixed import listing

FoamoftheSea committed:
Updated model name

FoamoftheSea committed:
Merge branch 'main' into pvtv2

FoamoftheSea committed:
Force empty commit for PR refresh

FoamoftheSea commented:
@amyeroberts I merged main today since the main branch passed CI tests, so I'm not sure why these checks are failing... the logs are not clear.

ArthurZucker commented:
The failing torch test is unrelated however the check repo consistency is related to bad `# Copied from`  statements. I don't have more tips but check the tabbing and if it's only wrapping a block or not!

FoamoftheSea commented:
Ah, good catch, thanks!. I will try to run a make fixup on that and see if I can get to the bottom of it.

FoamoftheSea committed:
Merge branch 'main' into pvtv2

FoamoftheSea committed:
Fixed linting issue

FoamoftheSea committed:
Removed # Copied from comments

FoamoftheSea committed:
Added PVTv2 to README_fr.md

FoamoftheSea committed:
Merge branch 'main' into pvtv2-staging

FoamoftheSea committed:
Ran make fix-copies

FoamoftheSea committed:
Merge branch 'main' into pvtv2-staging

FoamoftheSea committed:
Replace all FoamoftheSea hub references with OpenGVLab

FoamoftheSea committed:
Merge branch 'main' into pvtv2-staging

FoamoftheSea commented:
Looks like CIs are green ✔️ I had to remove all of the `# Copied from` comments from test_modeling_pvt_v2.py to get the checks to pass. None of the code blocks were really direct copies, and the only chunks of code that were were the smaller methods within larger test classes, which I guess can't be marked with this comment, but I haven't had a chance to dig into exactly how those checks work. In any case, here are the commits where I added/removed the `# Copied from` comments in case anyone wants to analyze best practice here any further.
- Added: https://github.com/huggingface/transformers/pull/26812/commits/fdb7ceb0e7646c6e263f7eee011b46f45344595f
- Removed: https://github.com/huggingface/transformers/pull/26812/commits/46f434efb3044ecac8e17abb6a5bed15f37c5340

The authors have reached out to me via email to let me know they've copied the pretrained weights over to the OpenGVLab HuggingFace account here: https://huggingface.co/collections/OpenGVLab/pvt-65db4ca6c3e37ebc67cd8e01, so I went ahead and switched all of the hub references in the code to these weights, and we should be ready to merge!

amyeroberts reviewed (approved):
Thanks for all the work on this model! 

Just a small comment about the config and a nit

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26812). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

FoamoftheSea commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,167 @@
+    ):
+        super().__init__(**kwargs)
+
+        if kwargs.get("_out_indices", None) is not None:
+            out_indices = kwargs["_out_indices"]
+            out_features = None
```
So the explanation for this is that without either `out_indices` or `out_features` not being set to `None`, the call to `get_aligned_output_features_output_indices` on line 165 will end up overriding whatever is in the pretrained config with a single value representing the last stage, leading to loading issues of pretrained models expecting these values to be consistent.

However, this solution is not optimal because it will always favor the config over the user input, and we would prefer the user to have the option to override the config, so I've come up with a better solution, which will prioritize passing these values as 1) User setting if not None, 2) Config setting if available, 3) Default to None

I will have a commit soon to fix this

Note: I originally followed what I saw in the ResNet config, so this might still be a problem for loading models with ResNet backbones that had specific out_indices configured, we may want to test for that.

amyeroberts replied:
I don't think this is in the ResNetConfig? 

It should be tested, but I believe is covered in `BackboneTesterMixin`

amyeroberts replied:
Just to make sure we're on the same page - what do you mean by "overwrite the settings in the pretrained config"?

FoamoftheSea replied:
When loading a pretrained config using `PvtV2Config.from_pretrained()`, the call to `super.__init__()` correctly finds and loads the `self._out_indices` and `self._out_features` from the JSON, and these are then accessible using the `self.out_indices` and `self.out_features` properties from the `BackboneConfigMixin`. However, the subsequent call to the `get_aligned_output_features_output_indices` function, if passed `None` for both `out_features` and `out_indices` (which are the default values in the `__init__` function and therefore the values when using `from_pretrained`), will actually revert these back to referencing only the last index/feature layer. Thus, the PVTv2Config, and presumably the ResNet config, would effectively *ignore* the `_out_indices` and `_out_features` fields from the pretrained JSON config. 

The code you originally highlighted was a hack to make sure that if there was already a self._out_indices set, that one of these variables would be not None when it reached the function call, but that was not the right solution. I believe the correct solution is to prioritize the values being fed to the function call as I laid out above.

I assumed that calling this line in the ResNetConfig `__init__` function was essential for some reason, which is why I copied it over, but then it caused problems. I think it's intended to allow for flexibility in loading different architectures, but I'm really not sure tbh, though I would hesitate to remove it.

FoamoftheSea replied:
Actually on second thought, I can see that the call to the function is necessary for establishing a default behavior when the options are not passed, and also for aligning any customization that the user is trying to accomplish. The issue was that it did not account for previously loaded values from the JSON and was causing them to be overridden.

amyeroberts replied:
> The issue was that it did not account for previously loaded values from the JSON and was causing them to be overridden.

Sorry, I'm still slightly confused about the behaviour being described here. Is this the preprocessor_config.json you're talking about? 

In this case, it should be possible to load previous values from the config i.e.

```py
config = PvtV2Config(out_indices=(2, 3))
config.save_pretrained("test_config")

config = PvtV2Config.from_pretrained("test_config")
assert config.out_indices == [2, 3]
```

FoamoftheSea replied:
Yes, this is the behavior in question.

I've just run a test and the ResNet config works fine, it appears that there is a different saving behavior, where the ResNetConfig is saving these fields into the JSON as "out_indices" and "out_features", whereas the PvtV2Config is saving them with the underscore prefix as "_out_indices" and "_out_features", and this is where the divergent behavior stems from.

When I get time later this evening, I can dig into why `save_pretrained` has a different behavior on these fields in ResNetConfig vs PvtV2Config, because that is where the fix should probably be.

FoamoftheSea committed:
Fixed out_indices and out_features logic in configuration_pvt_v2.py

FoamoftheSea committed:
Made ImageNet weight conversion verification optional in convert_pvt_v2_to_pytorch.py

FoamoftheSea committed:
Ran code fixup

amyeroberts commented on src/transformers/models/pvt_v2/configuration_pvt_v2.py:
```diff
@@ -0,0 +1,165 @@
+        self.qkv_bias = qkv_bias
+        self.linear_attention = linear_attention
+        self.stage_names = [f"stage{idx}" for idx in range(1, len(depths) + 1)]
+        self._out_features, self._out_indices = get_aligned_output_features_output_indices(
+            out_features=out_features if out_features is not None else getattr(self, "out_features", None),
+            out_indices=out_indices if out_indices is not None else getattr(self, "out_indices", None),
+            stage_names=self.stage_names,
+        )
```
I don't see why we need this here - neither `out_features` not `out_indices` should be already set when initializing the config 

```suggestion
        self._out_features, self._out_indices = get_aligned_output_features_output_indices(
            out_features=out_features,
            out_indices=out_indices,
            stage_names=self.stage_names,
        )
```

FoamoftheSea replied:
Ahhh, ok I see what happened. This was a tricky one. The order of the base classes matters here, as before we had `class ResNetConfig(BackboneConfigMixin, PretrainedConfig)` but `class PvtV2Config(PretrainedConfig, BackboneConfigMixin)`, and thus the latter was not getting the proper `to_dict` method override, and therefore not saving these fields correctly to JSON, which led to the weirdness during the `__init__` call

I have fixed this the correct way now in: https://github.com/huggingface/transformers/pull/26812/commits/0e7a0545296546f33fc3e9e1bbf0f2e78ae43e20

Thank you for your scrutiny here, I hadn't realized these values were not being saved properly. For a bit of backstory, I built the PvtV2 and was using it *before* I incorporated the Backbone mixins, so the code you highlighted was a hack that was part of the original work and prevented me from seeing this issue.

amyeroberts replied:
Thanks for digging into this. Adding on my to-do list to make sure this is easier to implement / catch

amyeroberts reviewed (commented):
Just a final comment on the config

FoamoftheSea committed:
Fixed order of parent classes in PvtV2Config to fix the to_dict method override

amyeroberts reviewed (approved):
Thanks for all the work adding this model - looks great! 💪
